### PR TITLE
add company user

### PR DESCRIPTION
**Pull request by Cheephot**

### CODE REVIEW

File: src/CompanyService.Broker/Consumers/CreateCompanyUserConsumer.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+namespace LT.DigitalOffice.CompanyService.Broker.Consumers
+{
+  public class CreateCompanyUserConsumer : IConsumer<ICreateCompanyUserRequest>
+  {
+    private readonly ICompanyUserRepository _companyUserRepository;
+    private readonly IDbCompanyUserMapper _companyUserMapper;
+
+    private async Task<bool> CreateAsync(ICreateCompanyUserRequest request)
+    {
+      await Task.WhenAll(
```
Comment (MariyaSin) [inline src/CompanyService.Broker/Consumers/CreateCompanyUserConsumer.cs:17]: no need use Task.WhenAll if you await only one response

File: src/CompanyService.Broker/Consumers/CreateCompanyUserConsumer.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
+﻿using System.Threading.Tasks;
+using LT.DigitalOffice.CompanyService.Data.Interfaces;
+using LT.DigitalOffice.CompanyService.Mappers.Db.Interfaces;
+using LT.DigitalOffice.Kernel.BrokerSupport.Broker;
+using LT.DigitalOffice.Models.Broker.Requests.Company;
+using MassTransit;
+
+namespace LT.DigitalOffice.CompanyService.Broker.Consumers
+{
+  public class CreateCompanyUserConsumer : IConsumer<ICreateCompanyUserRequest>
```
Comment (MariyaSin) [inline src/CompanyService.Broker/Consumers/CreateCompanyUserConsumer.cs:10]: CreateCompanyUsersConsumer

File: src/CompanyService.Broker/Consumers/DisactivateUserConsumer.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    public DisactivateUserConsumer(
+      ICompanyUserRepository companyUserRepository)
+    {
+      _companyUserRepository = companyUserRepository;
+    }
+
+    public async Task Consume(ConsumeContext<IDisactivateUserRequest> context)
+    {
+      await Task.WhenAll(
+        _companyUserRepository.RemoveAsync(context.Message.UserId, context.Message.ModifiedBy));
```
Comment (MariyaSin) [inline src/CompanyService.Broker/Consumers/DisactivateUserConsumer.cs:21]: no need use whenAll

File: src/CompanyService.Broker/Consumers/GetCompanyConsumer.cs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  {
+    private readonly ICompanyUserRepository _userRepository;
+    private readonly IOptions<RedisConfig> _redisConfig;
+    private readonly IRedisHelper _redisHelper;
+    private readonly ICacheNotebook _cacheNotebook;
+    private readonly ICompanyDataMapper _companyDataMapper;
+
+    private async Task<List<CompanyData>> GetCompanyAsync(IGetCompaniesRequest request)
+    {
+      List<DbCompanyUser> usersInfo = await _userRepository.GetAsync(request);
```
Comment (MariyaSin) [inline src/CompanyService.Broker/Consumers/GetCompanyConsumer.cs:31]: rename to companysUsers

File: src/CompanyService.Broker/Consumers/GetCompanyConsumer.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+      List<DbCompany> companies = usersInfo.Select(u => u.Company).Distinct().ToList();
```
Comment (MariyaSin) [inline src/CompanyService.Broker/Consumers/GetCompanyConsumer.cs:33]: Why do you not use include to get companyInfo with companyUsers?

File: src/CompanyService.Data.Provider.MsSql.Ef/Migrations/20211125172500_AddCompaniesUsersTable.cs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    protected override void Up(MigrationBuilder migrationBuilder)
+    {
+      migrationBuilder.CreateTable(
+        name: DbCompanyUser.TableName,
+        columns: table => new
+        {
+          Id = table.Column<Guid>(nullable: false),
+          UserId = table.Column<Guid>(nullable: false),
+          CompanyId = table.Column<Guid>(nullable: false),
+          Rate = table.Column<double>(nullable: false),
```
Comment (MariyaSin) [inline src/CompanyService.Data.Provider.MsSql.Ef/Migrations/20211125172500_AddCompaniesUsersTable.cs:21]: nullable true

File: src/CompanyService.Data.Provider.MsSql.Ef/Migrations/20211125172500_AddCompaniesUsersTable.cs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+          StartWorkingAt = table.Column<DateTime>(nullable: false),
```
Comment (MariyaSin) [inline src/CompanyService.Data.Provider.MsSql.Ef/Migrations/20211125172500_AddCompaniesUsersTable.cs:22]: nullable true

File: src/CompanyService.Data/CompanyUserRepository.cs [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    public async Task<bool> EditAsync(EditCompanyUserRequest request)
+    {
+      if (request is null)
+      {
+        return false;
+      }
+
+      var companyUser = await _provider.CompaniesUsers.FirstOrDefaultAsync(r => r.UserId == request.UserId);
```
Comment (MariyaSin) [inline src/CompanyService.Data/CompanyUserRepository.cs:54]: do not use var, check if companyUser is null

File: src/CompanyService.Data/CompanyUserRepository.cs [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+      return dbCompanyUser.Id;
+    }
+
+    public async Task<bool> DoesExistAsync(Guid userId)
+    {
+      return await _provider.CompaniesUsers.AnyAsync(u => u.UserId == userId);
+    }
+
+    public async Task<bool> EditAsync(EditCompanyUserRequest request)
```
Comment (MariyaSin) [inline src/CompanyService.Data/CompanyUserRepository.cs:47]: why you do not use patch?

File: src/CompanyService.Data/CompanyUserRepository.cs [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+    public async Task<List<DbCompanyUser>> GetAsync(List<Guid> userIds)
+    {
+      return await _provider.CompaniesUsers
+        .Include(cu => cu.Company)
+        .Where(u => userIds.Contains(u.UserId) && u.IsActive)
+        .ToListAsync();
+    }
+
+    public async Task RemoveAsync(Guid userId, Guid removedBy)
```
Comment (MariyaSin) [inline src/CompanyService.Data/CompanyUserRepository.cs:90]: no need take removed by from command

File: src/CompanyService.Mappers/Db/DbCompanyUserMapper.cs [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+  public class DbCompanyUserMapper : IDbCompanyUserMapper
+  {
+    public DbCompanyUser Map(ICreateCompanyUserRequest request)
+    {
+      return new DbCompanyUser
+      {
+        Id = Guid.NewGuid(),
+        CompanyId = request.CompanyId,
+        UserId = request.UserId,
+        Rate = request.Rate != null ? request.Rate : null,
```
Comment (MariyaSin) [inline src/CompanyService.Mappers/Db/DbCompanyUserMapper.cs:17]: ?

File: src/CompanyService.Mappers/Db/DbCompanyUserMapper.cs [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+        StartWorkingAt = request.StartWorkingAt != null ? request.StartWorkingAt : null,
```
Comment (MariyaSin) [inline src/CompanyService.Mappers/Db/DbCompanyUserMapper.cs:18]: what for do you do check?

File: src/CompanyService.Models.Dto/Requests/CompanyUser/EditCompanyUserRequest.cs [hunk 13]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+
+namespace LT.DigitalOffice.CompanyService.Models.Dto.Requests.CompanyUser
+{
+  public class EditCompanyUserRequest
```
Comment (MariyaSin) [inline src/CompanyService.Models.Dto/Requests/CompanyUser/EditCompanyUserRequest.cs:5]: record

File: src/CompanyService.Validation/CompanyUser/EditUserRateRequestValidator.cs [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace LT.DigitalOffice.CompanyService.Validation.CompanyUser
+{
+  public class EditUserRateRequestValidator : AbstractValidator<EditCompanyUserRequest>, IEditCompanyUserRequestValidator
+  {
+    public EditUserRateRequestValidator(
+      ICompanyUserRepository userRepository)
+    {
+      RuleFor(x => x.Rate)
+        .LessThanOrEqualTo(1)
```
Comment (MariyaSin) [inline src/CompanyService.Validation/CompanyUser/EditUserRateRequestValidator.cs:14]: add error messages

File: src/CompanyService.Validation/CompanyUser/EditUserRateRequestValidator.cs [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+        .GreaterThan(0);
+
+      RuleFor(x => x.UserId)
```
Comment (MariyaSin) [inline src/CompanyService.Validation/CompanyUser/EditUserRateRequestValidator.cs:17]: no need do it here, you can check it in edit repo method

File: src/CompanyService.Broker/Consumers/CreateCompanyUsersConsumer.cs [hunk 16]
```diff
@@ -0,0 +1,37 @@
+﻿using System.Threading.Tasks;
+using LT.DigitalOffice.CompanyService.Data.Interfaces;
+using LT.DigitalOffice.CompanyService.Mappers.Db.Interfaces;
+using LT.DigitalOffice.Kernel.BrokerSupport.Broker;
+using LT.DigitalOffice.Models.Broker.Requests.Company;
+using MassTransit;
+
+namespace LT.DigitalOffice.CompanyService.Broker.Consumers
+{
+  public class CreateCompanyUsersConsumer : IConsumer<ICreateCompanyUserRequest>
```
Comment (MariyaSin) [inline src/CompanyService.Broker/Consumers/CreateCompanyUsersConsumer.cs:10]: CreateCompanyUserConsumer

File: src/CompanyService.Broker/Consumers/GetCompanyConsumer.cs [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+using LT.DigitalOffice.Kernel.RedisSupport.Helpers.Interfaces;
+using LT.DigitalOffice.Models.Broker.Models.Company;
+using LT.DigitalOffice.Models.Broker.Requests.Company;
+using LT.DigitalOffice.Models.Broker.Responses.Company;
+using MassTransit;
+using Microsoft.Extensions.Options;
+
+namespace LT.DigitalOffice.CompanyService.Broker.Consumers
+{
+  public class GetCompanyConsumer : IConsumer<IGetCompaniesRequest>
```
Comment (MariyaSin) [inline src/CompanyService.Broker/Consumers/GetCompanyConsumer.cs:21]: GetCompaniesConsumer

File: src/CompanyService.Data/CompanyRepository.cs [hunk 18]
```diff
@@ -44,6 +47,20 @@ public async Task<DbCompany> GetAsync()
       return await _provider.Companies.FirstOrDefaultAsync();
     }
 
+    public async Task<List<DbCompany>> GetAsync(IGetCompaniesRequest request)
+    {
+      IQueryable<DbCompany> dbDepartments = _provider.Companies.AsQueryable();
+
+      if (request.UsersIds is not null && request.UsersIds.Any())
+      {
+        dbDepartments = dbDepartments.Where(d => d.Users.Any(du => request.UsersIds.Contains(du.UserId)));
```
Comment (MariyaSin) [inline src/CompanyService.Data/CompanyRepository.cs:56]: user must be IsActive

File: src/CompanyService.Models.Dto/Configuration/RabbitMqConfig.cs [hunk 19]
```diff
@@ -14,6 +14,9 @@ namespace LT.DigitalOffice.CompanyService.Models.Dto.Configuration
   public class RabbitMqConfig : BaseRabbitMqConfig
   {
     public string GetSmtpCredentialsEndpoint { get; set; }
+    public string CreateCompanyUserEndpoint { get; set; }
+    public string GetCompanyEndpoint { get; set; }
```
Comment (MariyaSin) [inline src/CompanyService.Models.Dto/Configuration/RabbitMqConfig.cs:18]: GetCompaniesEndpoint

File: src/CompanyService.Validation/CompanyUser/EditCompanyUserRequestValidator.cs [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+      AddСorrectPaths(
+        new()
+        {
+          nameof(EditCompanyUserRequest.Rate),
+          nameof(EditCompanyUserRequest.StartWorkingAt)
+        });
+
+      AddСorrectOperations(nameof(EditCompanyUserRequest.Rate), new() { OperationType.Replace });
+      AddСorrectOperations(nameof(EditCompanyUserRequest.StartWorkingAt), new() { OperationType.Replace });
+
```
Comment (MariyaSin) [inline src/CompanyService.Validation/CompanyUser/EditCompanyUserRequestValidator.cs:34]: remove empty line

File: src/CompanyService.Validation/CompanyUser/EditCompanyUserRequestValidator.cs [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+      #region StartWorkingAt
+
+      AddFailureForPropertyIf(
+        nameof(EditCompanyUserRequest.StartWorkingAt),
+        x => x == OperationType.Replace,
+        new Dictionary<Func<Operation<EditCompanyUserRequest>, bool>, string>
+        {
+          {
+            x => string.IsNullOrEmpty(x.value?.ToString())? true :
```
Comment (MariyaSin) [inline src/CompanyService.Validation/CompanyUser/EditCompanyUserRequestValidator.cs:46]: x.value?.ToString() is null ?

File: src/CompanyService.Validation/CompanyUser/EditCompanyUserRequestValidator.cs [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+      #endregion
+
+      #region Rate
+
+      AddFailureForPropertyIf(
+        nameof(EditCompanyUserRequest.Rate),
+        x => x == OperationType.Replace,
+        new()
+        {
+          { x => double.TryParse(x.value.ToString(), out double rate) && rate > 0 && rate <= 1,
```
Comment (MariyaSin) [inline src/CompanyService.Validation/CompanyUser/EditCompanyUserRequestValidator.cs:61]: x.vlue?.ToString() is null ? true : double.TryParse(x.value.ToString(), out double rate) && rate > 0 && rate <= 1

File: src/CompanyService/appsettings.json [hunk 23]
```diff
@@ -53,6 +53,8 @@
 
     // CompanyService receiving endpoints
     "GetSmtpCredentialsEndpoint": "CompanyService_GetSmtpCredentials",
+    "CreateCompanyUserEndpoint": "CompanyService_CreateCompanyUser",
+    "GetCompanyEndpoint": "CompanyService_GetCompany",
```
Comment (MariyaSin) [inline src/CompanyService/appsettings.json:57]: companies

**Review by MariyaSin — APPROVED** [review]

**Review by grinakate — APPROVED** [review]

**Review by irlirla — APPROVED** [review]

**Review by asyagraf — APPROVED** [review]

**Pull request merged**